### PR TITLE
#1223 DDC-3453 - make `EntityManager` constructor `public`

### PR DESCRIPTION
**Pull request by Ocramius**

### CODE REVIEW

**Comment by Ocramius** [comment]
Linking #1223

**Comment by stof** [comment]
@Ocramius If we make it public, some of the validation done in create probably needs to be moved in the constructor (checking that the connection uses the same EventManager for instance)

**Comment by beberlei** [comment]
Why?

**Comment by Ocramius** [comment]
> checking that the connection uses the same EventManager for instance

I don't think that is a responsibility of the EntityManager: I use multiple EntityManager instances sharing the same `Connection` object without problems.

 > Why?

The EntityManager is just painful to integrate in libraries with DICs. The `public static create()` was most probably introduced to simplify standalone usage of the ORM, but integration with more complex tools is just annoying as heck right now

**Comment by stof** [comment]
> I don't think that is a responsibility of the EntityManager: I use multiple EntityManager instances sharing the same Connection object without problems.

But are they using different EventManager instances or no ? If yes, I think this would create issues for the schema-related events

**Comment by stof** [comment]
@Ocramius decent DICs are able to use static factory methods AFAIK

**Comment by Ocramius** [comment]
>  If yes, I think this would create issues for the schema-related events

Yes, that does indeed happen, but it doesn't affect my own environment (no schema usage in my case, just sharing the same connection across different insulated EMs)

 > decent DICs are able to use static factory methods AFAIK

Yes, but that still makes it quite problematic to integrate. This approach is much simpler and doesn't seem to harm

**Comment by fprochazka** [comment]
> some of the validation done in create probably needs to be moved in the constructor

:+1: 

---

What is the point behind static factory anyway? Why is allowing public constructor a problem?

**Comment by guiwoda** [comment]
I'm having a problematic experience integrating the `EntityManager` in Laravel. More precisely:
- Laravel suggests constructor injection throughout all your classes.
- Hinting `EntityManager` and binding a closure that calls `create(...)` works, BUT
- Some objects in Laravel are created on all requests (through `ServiceProviders`), and as `EntityManager` requires a `Connection` object on construction, that makes all my requests require a valid DB connection.

This can probably be fixed on any of both sides. My fix was to use an `EntityManagerFactory` and pospone creation of the `EntityManager` to when it's actually needed. I also enforce a single instance through it.

Does this happen in other scenarios?

**Comment by mcspronko** [comment]
With respect to libraries with DIC support, we can configure EntityManager instance as shared in DIC configuration so only one EntityManager will be used across the system.

**Comment by Ocramius** [comment]
@guiwoda doesn't laravel support proxying services?

**Comment by guiwoda** [comment]
@Ocramius 

**TL;DR**: Artisan Commands are the places where depending on `EntityManager` would be a problem in Laravel.

You can defer loading of a ServiceProvider to when it actually needs the class. There's no such thing as `Services` in Laravel out-of-the-box, but if we define service as "a probably single-instanced object that will be consumed through the IoC container", then yes, you can do whatever you want with those. It's your binding to make. I usually declare my dependencies through constructor injection and map any interfaces to concrete classes on ServiceProvider, and reduce the traversal services down to a minimum.

The problem I usually face is with the `artisan` cli  tool:

In Laravel, you can run `php artisan` to list the available Commands associated in the project. This association is probably done through a ServiceProvider somewhere, and cannot be deferred.
When you do this, Laravel instantiates all associated Commands.

Problem is, commands receive deps through construction as any other object would, so any command that would use EntityManager would be instantiated (but not run) on every request and cli call.

This would also be solved if Command and CommandHandling would be separated. Or if the `php artisan` list of commands wouldn't need to instantiate them. Decoupling the command from the handling is relatively trivial, but it's not the default way to go with Laravel.

**Comment by Ocramius** [comment]
> Problem is, commands receive deps through construction as any other object would, so any command that would use EntityManager would be instantiated (but not run) on every request and cli call.

That's why I asked about proxying:

 - http://ocramius.github.io/blog/zf2-and-symfony-service-proxies-with-doctrine-proxies/
 - http://symfony.com/doc/current/components/dependency_injection/lazy_services.html
 - http://framework.zend.com/manual/current/en/modules/zend.service-manager.lazy-services.html

**Comment by guiwoda** [comment]
@Ocramius Read through your article just now. Totally right about it!
I'll check if there's something already done for Laravel on the DI container and proxies!

Thanks again!

**Comment by Ocramius** [comment]
@guiwoda poke me on IRC if you need any help with the implementation of that stuff.

**Comment by Ocramius** [comment]
After more careful analysis of this, I think that making the constructor `public` with a class with so many responsibilities is a bad idea for now: we'll likely add more named constructors (static factory methods) for 3.x, and therefore I am withdrawing this patch.
